Add rrcf outlier

Tatsuya-hasegawa opened the pull request:
Hello Ian,

I implemented Robust Random Cut Forest class and the outlier function.
RRCF is more useful when we detect anomaly of time series data as you know.

```
imported Robust Random Cut Forest python library
MIT License Copyright (c) 2018 kLabUM
https://klabum.github.io/rrcf/
```

Best regards,
Tatsuya

ianhelle commented on msticpy/analysis/outliers.py:
```diff
@@ -92,9 +125,226 @@ def identify_outliers(
+        """
+        Initialize Robust Random Cut Forest model.
+
+        Parameters
```
A minor nit. The docstring formats you're using are a bit inconsistent.
We try to stick to numpy format - which is a bit verbose but I think one of the clearest.

Tatsuya-hasegawa replied:
Sorry fot that, I changed to use numpy format for all the new functions.

ianhelle commented on msticpy/analysis/outliers.py:
```diff
@@ -92,9 +125,226 @@ def identify_outliers(
+    def decision_function(self, x: np.ndarray) -> np.ndarray:
+        """Compute anomaly scores for input data."""
+        # Score aggrigation in tree by parallel job.
+        print("Warning: RRCF decision_function may be too slow so takes long time.")
```
Might be better to use logging here rather than print.
maybe at INFO level.
This way people who are using it non-interactively can shut off unwanted output easily.
This example shows the basic pattern we use:
https://github.com/microsoft/msticpy/blob/8bc4652f016186325ee7fcbbc381850d2fe01239/msticpy/auth/azure_auth_core.py#L45
https://github.com/microsoft/msticpy/blob/8bc4652f016186325ee7fcbbc381850d2fe01239/msticpy/auth/azure_auth_core.py#L145

Tatsuya-hasegawa replied:
I changed to use logger. Thank you.

ianhelle commented on msticpy/analysis/outliers.py:
```diff
@@ -92,9 +125,226 @@ def identify_outliers(
+    MIT License Copyright (c) 2018 kLabUM.
+    (https://klabum.github.io/rrcf/).
+
+    Arguments:
```
heading should be Parameters (not Arguments)
Also the indentation of the string "Input data" is too much - this can cause the doc generation to get messed up.

ianhelle commented on msticpy/analysis/outliers.py:
```diff
@@ -42,6 +62,7 @@ def identify_outliers(
     x_predict: np.ndarray,
     contamination: float = 0.05,
     max_features: Optional[Union[int, float]] = None,
+    max_samples: Optional[Union[int, float]] = None,
```
Since you're editing this, wonder if you'd mind switching to use the newer, more use friendly annotation format?
I know the original use the typing types - so ignore this if it's too much to ask.
The pattern is 
```python
from __future__ import annotations
...
def identify_outliers(
    x: np.ndarray,
    x_predict: np.ndarray,
    contamination: float = 0.05,
    max_features: int | float | None = None,
    max_samples: int | float | None = None,
```

Also, instead of using the typing library forms like List[str] it can use builtin types like `list[str]`, `tuple[float, float]`, `dict[str, int]`
Florian is making his way through the code changing all of this stuff (and I do when I edit a file) so any help you can give us to modernize the code would be appreciated. :-)

Tatsuya-hasegawa replied:
Ok, I modified all of the annotations in outliers.py.

ianhelle reviewed (changes requested):
I'll have to trust your expertise on how this works and is implemented (which I do).
You might want to include one of the notebooks and add it to the notebooks folder.

I have a couple of lightweight comments - more about format than functionality.

ianhelle commented on msticpy/analysis/outliers.py:
```diff
@@ -32,6 +42,16 @@
 
+try:
+    import rrcf
+except ImportError as imp_err:
```
You might consider adding rrcf to the setup extra for ml
https://github.com/microsoft/msticpy/blob/8bc4652f016186325ee7fcbbc381850d2fe01239/setup.py#L47
Especially since you are indicating that this would be installed using `pip install msticpy[ml]`

Tatsuya-hasegawa commented:
Hi , Thanks for your many advices.
I think I was able to modify all the points from you except for including one of the notebooks and add it to the notebooks folder.
I will do it on another PR next time when I have get time.
Best regards,

ianhelle commented on msticpy/analysis/outliers.py:
```diff
@@ -11,13 +11,25 @@
 
 import numpy as np
 import pandas as pd
+from joblib import Parallel, delayed
```
Maybe need this in the "ml" extra as well

ianhelle commented:
Hi Tatsuya,
There are a bunch of typing warnings from mypy for outliers. In many cases it maybe as simple as declaring params as `int | float`.
Use:
```
from __future__ import annotations
```
at the top of the imports to ensure that Py3.8 supports this syntax.

Also - not a hard requirement - if you have any documentation to add about this (now impressive) module, it would def make it more visible to others. Also if you want to publish an article somewhere (we have a msticpy blog that hasn't been used for ages), we can include that in the release and linkedin/X posts.
Or just include the notebooks in the PR.
Great stuff, thx for the submission.

Tatsuya-hasegawa commented:
Hi Ian, 

Thank you for the advices.
I fixed the typing warning and my local check was fine.
FYI, I accidentally did a force push, but I changed it back, pulled, and then pushed again.

```
(base) hacket@hackeTlab msticpy % git commit -m "fixed CI/CD errors"  
check yaml...........................................(no files to check)Skipped
check json...........................................(no files to check)Skipped
trim trailing whitespace.................................................Passed
black....................................................................Passed
pylint...................................................................Passed
flake8...................................................................Passed
isort (python)...........................................................Passed
pydocstyle...............................................................Passed
ruff.....................................................................Passed
check_reqs_all...........................................................Passed
[add_rrcf_outlier c5037219] fixed CI/CD errors
 1 file changed, 19 insertions(+), 16 deletions(-)
 ```
 
 On the other hand,  I'm not sure for the compatibility to Python3.8...
 After some research with AI, I found that the "type OR operator (short for Union)" like "int | str" that msticpy is trying to migrate is a syntax introduced in Python 3.10 and later, and it doesn't seem to be supported in 3.8 even if you use from __future__ import annotations.
 So, please run CI/CD workflow again to check it.
 
 Also, about notebook documentation and artcle, I'm preparing some notebook files both the current IsolationForest outlier and this RRCF outliers.  I'm examining the results of differences for the same datasets.
 
 For time series data, Isolation Forest is known to be effective at detecting simple spikes, while RRCF is effective at detecting trend changes and correlations in multidimensional features. For non-time series data, Isolation Forest is also found to be faster and more accurate.

Tatsuya-hasegawa commented:
OMG, I'm sorry, 6 typing errors are still existing. I'll fix it day after tomorrow with also adding jupyter notebook documents.

In addition, I'll add the joblib package to pip requirement as same as rrcf package.
Missing packages:
 joblib:analysis.outliers.py

ianhelle commented:
No worries, take your time. :-)

Tatsuya-hasegawa commented:
Hi Ian,

I have finished my tasks  with adding some jupyter notebooks!

Best regards,


```
(base) hacket@hackeTlab msticpy % git commit -m "fixed the rest typing errors and add jupyter notebooks for comparing the outlier algorisms"
check yaml...........................................(no files to check)Skipped
check json...............................................................Passed
trim trailing whitespace.................................................Passed
black....................................................................Passed
pylint...................................................................Passed
flake8...................................................................Passed
isort (python)...........................................................Passed
pydocstyle...............................................................Passed
ruff.....................................................................Passed
check_reqs_all...........................................................Passed
[add_rrcf_outlier 2d66adab] fixed the rest typing errors and add jupyter notebooks for comparing the outlier algorisms
 8 files changed, 5673 insertions(+), 6 deletions(-)
 create mode 100644 docs/notebooks/Outliers-IsolationForest.ipynb
 create mode 100644 docs/notebooks/Outliers-IsolationForest_timeseries.ipynb
 create mode 100644 docs/notebooks/Outliers-RobustRandomCutForest.ipynb
 create mode 100644 docs/notebooks/Outliers-RobustRandomCutForest_timeseries.ipynb
```

Tatsuya-hasegawa commented:
Weird... 
What's that? Many errors are probably out of scope.

ianhelle commented:
Yeah - looks like something changed with maybe a mypy update and possible a respx update.
I will checkout your PR and fix these
